File: test/codegen/golden/expected/url_shortener/app/routers/url_mappings.py
from fastapi import APIRouter, Depends, HTTPException, Response
from fastapi.responses import RedirectResponse
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_session
from app.schemas.url_mapping import (
    ShortenRequest,
    UrlMappingRead,
)
from app.services.url_mapping import UrlMappingService

router = APIRouter(tags=["url_mapping"])


@router.post("/shorten", status_code=201)
async def shorten(
    body: ShortenRequest,
    session: AsyncSession = Depends(get_session),
) -> None:
    svc = UrlMappingService(session)
    return await svc.shorten(body)

@router.get("/urls", status_code=200)
async def list_all(
    session: AsyncSession = Depends(get_session),
) -> list[UrlMappingRead]:
    svc = UrlMappingService(session)
    return await svc.list_all()

@router.get("/{code}", status_code=302)
async def resolve(
    code: str,
    session: AsyncSession = Depends(get_session),
) -> RedirectResponse:
    svc = UrlMappingService(session)
    url = await svc.resolve(code)
    return RedirectResponse(url=url, status_code=302)

@router.delete("/{code}", status_code=204)
async def delete(
    code: str,
    session: AsyncSession = Depends(get_session),
) -> Response:
    svc = UrlMappingService(session)
    deleted = await svc.delete(code)
    if not deleted:
        raise HTTPException(status_code=404, detail="not found")
    return Response(status_code=204)

File: fixtures/golden/codegen/ts/express/sqlite/url_shortener/src/routes/urlMappings.ts
import type { Express, NextFunction, Request, Response } from 'express';

import { NotFound } from '../middleware/error.js';
import { validateBody } from '../middleware/validate.js';
import * as service from '../services/urlMapping.js';
import {
  UrlMappingCreateSchema,
  ShortenRequestSchema,
} from '../schemas/urlMapping.js';

const wrap =
  (handler: (req: Request, res: Response) => Promise<void>) =>
  (req: Request, res: Response, next: NextFunction): void => {
    handler(req, res).catch(next);
  };

export const registerUrlMappingRoutes = (app: Express): void => {
  app.post(
    '/shorten',
    validateBody(ShortenRequestSchema),
    wrap(async (req, res) => {
      const body = req.body as service.ShortenRequest;
      await service.shorten(body);
      res.status(201).end();
    }),
  );

  app.get(
    '/urls',
    wrap(async (_req, res) => {
      const result = await service.listAll();
      res.status(200).json(result);
    }),
  );

  app.get(
    '/:code',
    wrap(async (req, res) => {
      const code = req.params['code'] ?? '';
      const result = await service.resolve(code);
      if (result === null) {
        throw NotFound();
      }
      res.redirect(302, result.url);
    }),
  );

  app.delete(
    '/:code',
    wrap(async (req, res) => {
      const code = req.params['code'] ?? '';
      const ok = await service.delete_(code);
      if (!ok) {
        throw NotFound();
      }
      res.status(204).end();
    }),
  );

};
